### PR TITLE
Implement Brave Publisher Ads

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/7298

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- See specification

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by tmancey** [comment]
PR will be broken down once an initial code review has completed, then we can code review the new PRs after. Thanks

File: vendor/bat-native-ads/src/bat/ads/internal/bundle.cc [hunk 1]
```diff
@@ -45,8 +47,8 @@ bool Bundle::UpdateFromCatalog(const Catalog& catalog) {
 
   auto callback = std::bind(&Bundle::OnStateSaved,
       this, bundle_state->catalog_id, bundle_state->catalog_version,
-      bundle_state->catalog_ping,
-      bundle_state->catalog_last_updated_timestamp_in_seconds, _1);
+          bundle_state->catalog_ping,
+              bundle_state->catalog_last_updated_timestamp_in_seconds, _1);
```
Comment (masparrow) [inline vendor/bat-native-ads/src/bat/ads/internal/bundle.cc]: Alignment - and the next change below.

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/bundle.cc]: Alignment is correct as 4 spaces if following on from the previous line

File: vendor/bat-native-ads/src/bat/ads/publisher_ad_info.cc [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  if (document.HasMember("confirmation_type")) {
+    std::string confirmation_type = document["confirmation_type"].GetString();
+    type = ConfirmationType(confirmation_type);
+  }
+
+  return SUCCESS;
+}
+
+void SaveToJson(JsonWriter* writer, const PublisherAdInfo& info) {
+  writer->StartObject();
```
Comment (masparrow) [inline vendor/bat-native-ads/src/bat/ads/publisher_ad_info.cc]: Check `writer` is valid

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/publisher_ad_info.cc]: fixed

**Comment by tmancey** [comment]
Closing PR as rebased to master and broken PR into smaller PRs

**Pull request closed without merging**